<a href="https://colab.research.google.com/github/sahil-bhatia-iitbhu/pytorch/blob/main/Pytorch%20in%20one%20hour%20v2/Pytorch%20in%20One%20Hour%20code%20v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PyTorch for LLMs in 1-Hour

## Introduction

PyTorch (https://pytorch.org/) is an open-source Python-based deep learning library.

One of the reasons why PyTorch is so popular is its user-friendly interface and efficiency.However, despite its accessibility, it doesn’t compromise on flexibility, providing advancedusers the ability to tweak lower-level aspects of their models for customization and optimization. In short, for many practitioners and researchers, PyTorch offers just the right balance between usability and features.

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/1.webp" width="500px">

PyTorch's three main components include a tensor library as a fundamental building block for computing, automatic differentiation for model optimization, and deep learning utility functions, makingit easier to implement and train deep neural network models.

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/2.webp" width="500px">

Deep learning is a subcategory of machine learning that is focused on the implementation ofdeep neural networks. In turn, machine learning is a subcategory of AI that is concerned withalgorithms that learn from data. AI is the broader concept of machines being able to perform tasks that typically require human intelligence.

## A.1 Setting Up PyTorch

**Python version :**

Many scientific computing libraries do not immediately support the newest version of Python. Therefore, when installing PyTorch, it’s advisable to use a version of Python that is one ortwo releases older. For instance, if the latest version of Python is 3.13, using Python 3.11 or 3.12 is recommended.

**CUDA Compatibility :**

To explicitly install the CUDA-compatible version of PyTorch, it’s often better to specify the CUDA you want PyTorch to be compatible with. PyTorch’s official website (https://pytorch.org) provides commands to install PyTorch with CUDA support for different operating systems as shown in figure below.

Do check the summary blog on "[How-to-Check-Which-CUDA-Version-Your-Graphics-Card-Supports](https://sahil-bhatia-iitbhu.notion.site/How-to-Check-Which-CUDA-Version-Your-Graphics-Card-Supports-and-Ensure-Compatibility-with-PyTorch-CU-284601ee8073803c9b0ed5c3fd77e24d?source=copy_link)"

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/4.webp" width="500px">

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/5.webp" width="500px">

In [ ]:
import subprocess
import torch
import re

def get_driver_version():
    try:
        output = subprocess.check_output(['nvidia-smi']).decode()
        match = re.search(r'CUDA Version: (\d+\.\d+)', output)
        if match:
            return float(match.group(1))
    except Exception as e:
        return f"Error getting driver version: {e}"

def get_toolkit_version():
    try:
        output = subprocess.check_output(['nvcc', '--version']).decode()
        match = re.search(r'release (\d+\.\d+)', output)
        if match:
            return float(match.group(1))
    except Exception as e:
        return f"Error getting toolkit version: {e}"

driver_version = get_driver_version()
toolkit_version = get_toolkit_version()
torch_cuda_version = torch.version.cuda
torch_cuda_is_available = torch.cuda.is_available()

print(f"CUDA Driver Version (nvidia-smi): {driver_version}")
print(f"CUDA Toolkit Version (nvcc): {toolkit_version}")
print(f"PyTorch CUDA Version: {torch_cuda_version}")
print(f"Is CUDA available for PyTorch? {torch_cuda_is_available}")

if isinstance(driver_version, float) and torch_cuda_version is not None:
    if driver_version < float(torch_cuda_version):
        print("\nWarning: CUDA driver version is older than PyTorch CUDA version. Downgrade your PyTorch CUDA version.")
    else:
        print("\nCUDA driver version is compatible with or newer than PyTorch CUDA version.")

CUDA Driver Version (nvidia-smi): 12.4
CUDA Toolkit Version (nvcc): 12.5
PyTorch CUDA Version: 12.4
Is CUDA available for PyTorch? True

CUDA driver version is compatible with or newer than PyTorch CUDA version.


In [ ]:
# Uninstall current PyTorch version if compatibility is an issue
!pip uninstall -y torch torchvision torchaudio
!pip3 install torch torchaudio torchvision --index-url https://download.pytorch.org/whl/cu124

import torch

In [ ]:
print(torch.__version__)
print(torch.cuda.is_available())

2.6.0+cu124
True


### PyTorch on Apple Silicon :
If you have an Apple Mac with an Apple Silicon chip (like the M1, M2,M3, M4 or newer models), you have the option to leverage its capabilities to accelerate PyTorch code execution. To use your Apple Silicon chip for PyTorch, you first need to install PyTorch as younormally would. Then, to check if your Mac supports PyTorch acceleration with its Apple Silicon chip,you can run a simple code snippet in Python:

```
print(torch.backends.mps.is_available())
```
 If it returns True, it means that your Mac has an Apple Silicon chip that can be used to acceleratePyTorch code.

## A.2 Understanding tensors

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/6.webp" width="400px">

From a computational perspective, tensors serve as data containers. For instance, they hold multi-dimensional data, where each dimension represents a different feature. Tensor libraries, such as PyTorch, can create, manipulate, and compute with these multi-dimensional arrays efficiently. In this context, a tensor library functions as an array library.

### A.2.1 Scalars, vectors, matrices, and tensors

In [ ]:
import torch
import numpy as np

# create a 0D tensor (scalar) from a Python integer
tensor0d = torch.tensor(1)

# create a 1D tensor (vector) from a Python list
tensor1d = torch.tensor([1, 2, 3])

# create a 2D tensor from a nested Python list
tensor2d = torch.tensor([[1, 2],
                         [3, 4]])

# create a 3D tensor from a nested Python list
tensor3d_1 = torch.tensor([[[1, 2], [3, 4]],
                           [[5, 6], [7, 8]]])

# create a 3D tensor from NumPy array
ary3d = np.array([[[1, 2], [3, 4]],
                  [[5, 6], [7, 8]]])
tensor3d_2 = torch.tensor(ary3d)  # Copies NumPy array
tensor3d_3 = torch.from_numpy(ary3d)  # Shares memory with NumPy array

In [ ]:
ary3d[0, 0, 0] = 999
print(tensor3d_2) # remains unchanged

tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])


In [ ]:
print(tensor3d_3) # changes because of memory sharing

tensor([[[999,   2],
         [  3,   4]],

        [[  5,   6],
         [  7,   8]]])


### A.2.2 Tensor data types

In [ ]:
tensor1d = torch.tensor([1, 2, 3])
print(tensor1d.dtype)

torch.int64


In [ ]:
floatvec = torch.tensor([1.0, 2.0, 3.0])
print(floatvec.dtype)

torch.float32


In [ ]:
floatvec = tensor1d.to(torch.float32)
print(floatvec.dtype)

torch.float32


### A.2.3 Common PyTorch tensor operations

For readers who’d like to browse through all the different tensor operations available in PyTorch (hint: we won’t need most of these), I recommend checking out the official documentation at
https://pytorch.org/docs/stable/tensors.html

In [ ]:
tensor2d = torch.tensor([[1, 2, 3],
                         [4, 5, 6]])
tensor2d

tensor([[1, 2, 3],
        [4, 5, 6]])

In [ ]:
tensor2d.shape

torch.Size([2, 3])

In [ ]:
tensor2d.reshape(3, 2)

tensor([[1, 2],
        [3, 4],
        [5, 6]])

In [ ]:
tensor2d.view(3, 2)

tensor([[1, 2],
        [3, 4],
        [5, 6]])

Similar to `.reshape` and `.view`, there are several cases where PyTorch offers multiple syntax options for executing the same computation. This is because PyTorch initially followed the original Lua Torch syntax convention but then also added syntax to make it more similar toNumPy upon popular request.

In [ ]:
tensor2d.T

tensor([[1, 4],
        [2, 5],
        [3, 6]])

In [ ]:
tensor2d.matmul(tensor2d.T)

tensor([[14, 32],
        [32, 77]])

In [ ]:
tensor2d.mm(tensor2d.T)

tensor([[14, 32],
        [32, 77]])

In [ ]:
tensor2d @ tensor2d.T

tensor([[14, 32],
        [32, 77]])

## A.3 Seeing models as computation graphs

PyTorch’s automatic differentiation engine, also known as autograd. PyTorch’s autograd system provides functions to compute gradients in dynamic computational graphs automatically.

A computational graph (or computation graph in short) is a directed graph that allows us to express and visualize mathematical expressions. In the context of deep learning, a computation graph lays out the sequence of calculations needed to compute the output of a neural network – we will need this later to compute the required gradients for backpropagation, which is the main training algorithm for neural networks.

Example as below : LOGISTIC REGRESSION with single hidden-layer

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/7.webp" width="600px">

A logistic regression forward pass as a computation graph. The input feature `x1` is multiplied by a model weight `w1` and passed through an activation function `sigmoid(σ)` after adding the bias. The loss is computed by comparing the model output `a` with a given label `y`.

In fact, PyTorch builds such a computation graph in the background, and we can use this to calculate gradients of a loss function with respect to the model parameters (here `w1` and `b`) to train the model, which is the topic of the upcoming sections.

In [ ]:
import torch.nn.functional as F

y = torch.tensor([1.0])  # true label
x1 = torch.tensor([1.1]) # input feature
w1 = torch.tensor([2.2]) # weight parameter
b = torch.tensor([0.0])  # bias unit

z = x1 * w1 + b          # net input
a = torch.sigmoid(z)     # activation & output
# a = torch.tanh(z)     # activation & output

loss = F.binary_cross_entropy(a, y)
print(loss)

tensor(0.0852)


## A.4 Automatic differentiation made easy

In the previous section, we introduced the concept of computation graphs. If we carry out computations in PyTorch, it will build such a graph internally by default if one of its terminal nodes has the `requires_grad` attribute set to `True`.
This is useful if we want to compute gradients. Gradients are required when training neural networks via the popularbackpropagation algorithm, which can be thought of as an implementation of the chain rule from calculus for neural networks, which is illustrated in Figure below.

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/8.webp" width="700px">

In [2]:
import torch
import torch.nn.functional as F
from torch.autograd import grad

y = torch.tensor([1.0])
x1 = torch.tensor([1.1])
w1 = torch.tensor([2.2], requires_grad=True)
b = torch.tensor([0.0], requires_grad=True)

z = x1 * w1 + b
a = torch.sigmoid(z)

loss = F.binary_cross_entropy(a, y)

grad_L_w1 = grad(loss, w1, retain_graph=True)
grad_L_b = grad(loss, b, retain_graph=True)

print(grad_L_w1)
print(grad_L_b)

(tensor([-0.0898]),)
(tensor([-0.0817]),)


By default, PyTorch destroys the computation graph after calculating the gradients to freememory. However, since we are going to reuse this computation graph shortly, we set `retain_graph=True` so that it stays in memory.

Above, we have been using the grad function “manually,” which can be useful for experimentation, debugging, and demonstrating concepts. But in practice, PyTorch provideseven more high-level tools to automate this process. For instance, we can call `.backward` on the loss, and PyTorch will compute the gradients of all the leaf nodes in the graph, which willbe stored via the tensors’ `.grad` attributes:

In [3]:
loss.backward()

print(w1.grad)
print(b.grad)

tensor([-0.0898])
tensor([-0.0817])


## A.5 Implementing multilayer neural networks

In the previous sections, we covered PyTorch’s tensor and autograd components. This section focuses on PyTorch as a library for implementing deep neural networks.
To provide a concrete example, we focus on a multilayer perceptron, which is a fullyconnected neural network, as illustrated in Figure below.

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/9.webp" width="500px">

When implementing a neural network in PyTorch, we typically subclass the
torch.nn.Module
class to define our own custom network architecture. This
Module
base class provides a lot of functionality, making it easier to build and train models. For instance, it allows us toencapsulate layers and operations and keep track of the model’s parameters.

Within this subclass, we define the network layers in the
`__init__`
constructor and specifyhow they interact in the
`forward`
method. The
`forward`
method describes how the input datapasses through the network and comes together as a computation graph.

In contrast, the backward method, which we typically do not need to implement ourselves, is used during training to compute gradients of the loss function with respect to the model parameters.

Note that we used the
`Sequential`
class when we implemented the
`NeuralNetwork`
class.Using
`Sequential`
is not required, but it can make our life easier if we have a series of layersthat we want to execute in a specific order, as is the case here. This way, after instantiating
`self.layers = Sequential(...)`
in the
`__init__`
constructor, we just have to call the
`self.layers`
instead of calling each layer individually in the
`NeuralNetwork`
’s forward method.

In [8]:
class NeuralNetwork(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs):
        super().__init__()

        self.layers = torch.nn.Sequential(

            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30),
            torch.nn.ReLU(),

            # 2nd hidden layer
            torch.nn.Linear(30, 20),
            torch.nn.ReLU(),

            # output layer
            torch.nn.Linear(20, num_outputs),
        )

    def forward(self, x):
        logits = self.layers(x)
        return logits

In [9]:
model = NeuralNetwork(50, 3)

But before using this new model object, it is often useful to call print on the model to see asummary of its structure:

In [10]:
print(model)

NeuralNetwork(
  (layers): Sequential(
    (0): Linear(in_features=50, out_features=30, bias=True)
    (1): ReLU()
    (2): Linear(in_features=30, out_features=20, bias=True)
    (3): ReLU()
    (4): Linear(in_features=20, out_features=3, bias=True)
  )
)


Next, let’s check the total number of trainable parameters of this model:

In [11]:
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total number of trainable model parameters:", num_params)

Total number of trainable model parameters: 2213


Note that each parameter for which
`requires_grad=True`
counts as a trainable parameterand will be updated during training.

In the case of our neural network model with the two hidden layers above, these trainable parameters are contained in the `torch.nn.Linear` layers. A
*linear*
layer multiplies the inputs with a weight matrix and adds a bias vector. This is sometimes also referred to as a
*feedforward*
or
*fully connected*
layer.

Based on the
`print(model)`
call we executed above, we can see that the first Linear layer isat index position 0 in the layers attribute. We can access the corresponding weightparameter matrix as follows:

In [12]:
print(model.layers[0].weight)

Parameter containing:
tensor([[-0.0125,  0.0925, -0.0668,  ...,  0.0374, -0.0103,  0.0835],
        [-0.0184,  0.1350,  0.1097,  ..., -0.0007, -0.0884, -0.1390],
        [ 0.0626,  0.0448, -0.0358,  ...,  0.0256, -0.0028, -0.1318],
        ...,
        [-0.0461,  0.0060, -0.0403,  ...,  0.1051,  0.0519, -0.0246],
        [-0.0562, -0.0490, -0.0710,  ..., -0.0898,  0.1206,  0.0589],
        [-0.0858,  0.0605, -0.1015,  ..., -0.0827,  0.1273,  0.0182]],
       requires_grad=True)


In [ ]:
print(model.layers[0].weight.shape)

torch.Size([30, 50])


Note that if you execute the code above on your computer, the numbers in the weight matrixwill likely differ from those shown above. This is because the model weights are initializedwith small random numbers, which are different each time we instantiate the network. Indeep learning, initializing model weights with small random numbers is desired to break symmetry during training – otherwise, the nodes would be just performing the sameoperations and updates during backpropagation, which would not allow the network to learncomplex mappings from inputs to outputs.

However, while we want to keep using small random numbers as initial values for our layer weights, we can make the random number initialization reproducible by seeding PyTorch’srandom number generator via
`manual_seed`:

In [13]:
torch.manual_seed(123)

model = NeuralNetwork(50, 3)
print(model.layers[0].weight)

Parameter containing:
tensor([[-0.0577,  0.0047, -0.0702,  ...,  0.0222,  0.1260,  0.0865],
        [ 0.0502,  0.0307,  0.0333,  ...,  0.0951,  0.1134, -0.0297],
        [ 0.1077, -0.1108,  0.0122,  ...,  0.0108, -0.1049, -0.1063],
        ...,
        [-0.0787,  0.1259,  0.0803,  ...,  0.1218,  0.1303, -0.1351],
        [ 0.1359,  0.0175, -0.0673,  ...,  0.0674,  0.0676,  0.1058],
        [ 0.0790,  0.1343, -0.0293,  ...,  0.0344, -0.0971, -0.0509]],
       requires_grad=True)


In [14]:
torch.manual_seed(123)

X = torch.rand((1, 50))
out = model(X)
print(out)

tensor([[-0.1262,  0.1080, -0.1792]], grad_fn=<AddmmBackward0>)


In the code above, we generated a single random training example
`X`
as a toy input (notethat our network expects 50-dimensional feature vectors) and fed it to the model, returningthree scores. When we call
`model(x)`
, it will automatically execute the forward pass of the model.

The forward pass refers to calculating output tensors from input tensors. This involvespassing the input data through all the neural network layers, starting from the input layer,through hidden layers, and finally to the output layer.

These three numbers returned above correspond to a score assigned to each of the threeoutput nodes. Notice that the output tensor also includes a
`grad_fn`
value.

Here,
`grad_fn=<AddmmBackward0>`
represents the last-used function to compute a variable inthe computational graph. In particular,
`grad_fn=<AddmmBackward0>`
means that the tensor weare inspecting was created via a matrix multiplication and addition operation. PyTorch will use this information when it computes gradients during backpropagation. The
`<AddmmBackward0>`
part of
`grad_fn=<AddmmBackward0>`
specifies the operation that was performed. In this case, it is an
`Addmm`
operation.
`Addmm`
stands for matrix multiplication (
`mm`
)followed by an addition (
`Add`
).

If we just want to use a network without training or backpropagation, for example, if we use itfor prediction after training, constructing this computational graph for backpropagation can bewasteful as it performs unnecessary computations and consumes additional memory. So,when we use a model for inference (for instance, making predictions) rather than training, itis a best practice to use the
torch.no_grad()
context manager, as shown below. This tellsPyTorch that it doesn’t need to keep track of the gradients, which can result in significant savings in memory and computation.

In [15]:
with torch.no_grad():
    out = model(X)
print(out)

tensor([[-0.1262,  0.1080, -0.1792]])


In PyTorch, it’s common practice to code models such that they return the outputs of the lastlayer (
`logits`
) without passing them to a nonlinear activation function. That’s because PyTorch’s commonly used loss functions combine the `softmax` (or `sigmoid` for binary classification) operation with the `negative log-likelihood loss` in a single class. The reason for this is numerical efficiency and stability. So, if we want to compute class-membership probabilities for our predictions, we have to call the `softmax` function explicitly:

In [16]:
with torch.no_grad():
    out = torch.softmax(model(X), dim=1)
print(out)

tensor([[0.3113, 0.3934, 0.2952]])


The values can now be interpreted as class-membership probabilities that sum up to 1. Thevalues are roughly equal for this random input, which is expected for a randomly initialized model without training.

## A.6 Setting up efficient data loaders

Before we can train thismodel, we have to briefly talk about creating efficient data loaders in PyTorch, which we williterate over when training the model. The overall idea behind data loading in PyTorch is illustrated in Figure below.

PyTorch implements a `Dataset` and a `DataLoader` class. The `Dataset` class is used toinstantiate objects that define how each data record is loaded. The `DataLoader` handles how the datais shuffled and assembled into batches.

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/10.webp" width="700px">

Let’s start by creating a simple toy dataset of five training examples with two features each.Accompanying the training examples, we also create a tensor containing the correspondingclass labels: three examples belong to class 0, and two examples belong to class 1. Inaddition, we also make a test set consisting of two entries. The code to create this dataset isshown below:

In [17]:
X_train = torch.tensor([
    [-1.2, 3.1],
    [-0.9, 2.9],
    [-0.5, 2.6],
    [2.3, -1.1],
    [2.7, -1.5]
])

y_train = torch.tensor([0, 0, 0, 1, 1])

In [18]:
X_test = torch.tensor([
    [-0.8, 2.8],
    [2.6, -1.6],
])

y_test = torch.tensor([0, 1])

**IMPORTANT NOTE**

`Class Label Numbering`

PyTorch requires that class labels start with label 0, and the largestclass label value should not exceed the number of output nodes minus 1 (since Python indexcounting starts at 0. So, if we have class labels 0, 1, 2, 3, and 4, the neural network outputlayer should consist of 5 nodes.

Next, we create a custom dataset class,
`ToyDataset`
, by subclassing from PyTorch’s
`Dataset`
parent class, as shown below:

In [19]:
from torch.utils.data import Dataset


class ToyDataset(Dataset):
    def __init__(self, X, y):
        self.features = X
        self.labels = y

    def __getitem__(self, index):
        one_x = self.features[index]
        one_y = self.labels[index]
        return one_x, one_y

    def __len__(self):
        return self.labels.shape[0]

train_ds = ToyDataset(X_train, y_train)
test_ds = ToyDataset(X_test, y_test)

This custom
`ToyDataset`
class’s purpose is to use it to instantiate a PyTorch
DataLoader
.

But before we get to this step, let’s briefly go over the general structure of the
`ToyDataset`
code.
In PyTorch, the three main components of a custom
`Dataset`
class are the
`__init__`
constructor, the
`__getitem__`
method, and the
`__len__`
method, as shown in code
`ToyDataset`
code above.

In the
`__init__`
method, we set up attributes that we can access later in the
`__getitem__`
and
`__len__`
methods. This could be file paths, file objects, database connectors, and so on.Since we created a tensor dataset that sits in memory, we are simply assigning
`X`
and
`y`
to these attributes, which are placeholders for our tensor objects.

In the
`__getitem__`
method, we define instructions for returning exactly one item from the dataset via an index. This means the features and the class label corresponding to a single training example or test instance. (The data loader will provide this index, which we will cover shortly.)

Finally, the
`__len__`
method contains instructions for retrieving the length of the dataset. Here, we use the `.shape` attribute of a tensor to return the number of rows in the feature array.

In the case of the training dataset, we have five rows, which we can double-check asfollows:

In [20]:
len(train_ds)

5

Now that we defined a PyTorch
`Dataset`
class we can use for our toy dataset, we can use PyTorch’s
`DataLoader`
class to sample from it, as shown in the code below:

In [21]:
from torch.utils.data import DataLoader

torch.manual_seed(123)

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0
)

In [22]:
test_ds = ToyDataset(X_test, y_test)

test_loader = DataLoader(
    dataset=test_ds,
    batch_size=2,
    shuffle=False,
    num_workers=0
)

After instantiating the training data loader, we can iterate over it as shown below. (The iteration over the
`test_loader`
works similarly but is omitted for brevity.)

In [23]:
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y)

Batch 1: tensor([[ 2.3000, -1.1000],
        [-0.9000,  2.9000]]) tensor([1, 0])
Batch 2: tensor([[-1.2000,  3.1000],
        [-0.5000,  2.6000]]) tensor([0, 0])
Batch 3: tensor([[ 2.7000, -1.5000]]) tensor([1])


As we can see based on the output above, the
train_loader
iterates over the trainingdataset visiting each training example exactly once. This is known as a training epoch. Sincewe seeded the random number generator using
`torch.manual_seed(123)`
above, you shouldget the exact same shuffling order of training examples as shown above. However if you iterate over the dataset a second time, you will see that the shuffling order will change. This is desired to prevent deep neural networks getting caught in repetitive update cycles during training.

Note that we specified a batch size of 2 above, but the 3rd batch only contains a singleexample. That’s because we have five training examples, which is not evenly divisible by 2. In practice, having a substantially smaller batch as the last batch in a training epoch candisturb the convergence during training. To prevent this, it’s recommended to set
`drop_last=True`
, which will drop the last batch in each epoch, as shown below:

In [24]:
train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0,
    drop_last=True
)

In [25]:
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y)

Batch 1: tensor([[-1.2000,  3.1000],
        [-0.5000,  2.6000]]) tensor([0, 0])
Batch 2: tensor([[ 2.3000, -1.1000],
        [-0.9000,  2.9000]]) tensor([1, 0])


Lastly, let’s discuss the setting
`num_workers=0`
in the
`DataLoader`
. This parameter in PyTorch’s
`DataLoader`
function is crucial for parallelizing data loading and preprocessing.

When
`num_workers`
is set to 0, the data loading will be done in the main process and not in separate worker processes. This might seem unproblematic, but it can lead to significant slowdowns during model training when we train larger networks on a GPU. This is because instead of focusing solely on the processing of the deep learning model, the CPU must also take time to load and preprocess the data. As a result, the GPU can sit idle while waiting for the CPU to finish these tasks.

In contrast, when
`num_workers`
is set to a number greater than zero, multiple worker processes are launched to load data in parallel, freeing the main process to focus on training your model and better utilizing your system’s resources, which is illustrated in Figure below.

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/11.webp" width="700px">

However, if we are working with very small datasets, setting
`num_workers`
to 1 or larger may not be necessary since the total training time takes only fractions of a second anyway.

On the contrary, if you are working with tiny datasets or interactive environments such as `Jupyter notebooks`, increasing
`num_workers`
may not provide any noticeable speedup. They might, infact, lead to some issues. **One potential issue is the overhead of spinning up multiple worker processes, which could take longer than the actual data loading when your dataset is small.**
**Furthermore, for `Jupyter notebooks`, setting
`num_workers`
to greater than 0 can sometimes lead to issues related to the sharing of resources between different processes, resulting in errors or notebook crashes.** Therefore, it’s essential to understand the trade-off and make a calculated decision on setting the
`num_workers`
parameter. When used correctly, it can be a beneficial tool but should be adapted to your specific dataset size and computational environment for optimal results.

In my experience, setting
`num_workers=4`
usually leads to optimal performance on many real-world datasets, but optimal settings depend on your hardware and the code used for loading a training example defined in the
`Dataset`
class.

## A.7 A typical training loop

So far, we’ve discussed all the requirements for training neural networks: PyTorch’s tensorlibrary, autograd, the
`Module`
API, and efficient data loaders. Let’s now combine all thesethings and train a neural network on the toy dataset from the previous section. The trainingcode is shown in code below.

In [26]:
import torch.nn.functional as F


torch.manual_seed(123)
model = NeuralNetwork(num_inputs=2, num_outputs=2)
optimizer = torch.optim.SGD(model.parameters(), lr=0.5)

num_epochs = 3

for epoch in range(num_epochs):

    model.train()
    for batch_idx, (features, labels) in enumerate(train_loader):

        logits = model(features)

        loss = F.cross_entropy(logits, labels) # Loss function

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        ### LOGGING
        print(f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
              f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
              f" | Train/Val Loss: {loss:.2f}")

    model.eval()
    # Optional model evaluation

Epoch: 001/003 | Batch 000/002 | Train/Val Loss: 0.75
Epoch: 001/003 | Batch 001/002 | Train/Val Loss: 0.65
Epoch: 002/003 | Batch 000/002 | Train/Val Loss: 0.44
Epoch: 002/003 | Batch 001/002 | Train/Val Loss: 0.13
Epoch: 003/003 | Batch 000/002 | Train/Val Loss: 0.03
Epoch: 003/003 | Batch 001/002 | Train/Val Loss: 0.00


As we can see, the loss reaches zero after 3 epochs, a sign that the model converged on the training set. However, before we evaluate the model’s predictions, let’s go over some of the details of the preceding code.

First, note that we initialized a model with two inputs and two outputs. That’s because the `toydataset` from the previous section has two input features and two class labels to predict. Weused a stochastic gradient descent (
`SGD`
) optimizer with a learning rate (
`lr`
) of 0.5. **The `learning rate` is a hyperparameter, meaning it’s a tunable setting that we have to experiment with based on observing the loss. Ideally, we want to choose a learning rate such that the loss converges after a certain number of epochs – the number of epochs is another hyperparameter to choose.**

In practice, we often use a third dataset, a so-called `validation` dataset, to find the optimal hyperparameter settings. A `validation` dataset is similar to a `test` set. However, while we only want to use a `test` set precisely once to avoid biasing the evaluation, we usually use the `validation` set multiple times to tweak the model settings.

We also introduced new settings called
`model.train()`
and
`model.eval()`
. As these names imply, these settings are used to put the model into a training and an evaluation mode. Thisis necessary for components that behave differently during training and inference, such as
`dropout`
or
`batch normalization`
layers. Since we don’t have dropout or other components in our
`NeuralNetwork`
class that are affected by these settings, using
`model.train()`
and
`model.eval()`
is redundant in our code above.

**However, it’s best practice to include them anyway to avoid unexpected behaviors when we change the model architecture or reuse thecode to train a different model.**

As discussed earlier, we pass the logits directly into the
`cross_entropy`
loss function, which will apply the `softmax` function internally for efficiency and numerical stability reasons. Then,calling
`loss.backward()`
will calculate the gradients in the computation graph that PyTorch constructed in the background. The
`optimizer.step()`
method will use the gradients to update the model parameters to minimize the loss.

In the case of the SGD optimizer, this means multiplying the gradients with the learning rate and adding the scaled negative gradient to the parameters.

**Preventing undesired gradient accumulation**

It is important to include an `optimizer.zero_grad()` call in each update round to reset the gradients to zero. Otherwise, the gradients will accumulate, which may be undesired.

In [27]:
model.eval()

with torch.no_grad():
    outputs = model(X_train)

print(outputs)

tensor([[ 2.8569, -4.1618],
        [ 2.5382, -3.7548],
        [ 2.0944, -3.1820],
        [-1.4814,  1.4816],
        [-1.7176,  1.7342]])


To obtain the class membership probabilities, we can then use PyTorch’s `softmax` function, as follows:

In [28]:
torch.set_printoptions(sci_mode=False)
probas = torch.softmax(outputs, dim=1)
print(probas)

predictions = torch.argmax(probas, dim=1)
print(predictions)

tensor([[    0.9991,     0.0009],
        [    0.9982,     0.0018],
        [    0.9949,     0.0051],
        [    0.0491,     0.9509],
        [    0.0307,     0.9693]])
tensor([0, 0, 0, 1, 1])


Let’s consider the first row in the code output above. Here, the first value (column) meansthat the training example has a 99.91% probability of belonging to class 0 and a 0.09%probability of belonging to class 1. (The
`set_printoptions`
call is used here to make theoutputs more legible.)

We can convert these values into class labels predictions using PyTorch’s
`argmax`
function, which returns the index position of the highest value in each row if we set
`dim=1`
(setting
`dim=0`
would return the highest value in each column, instead):

In [31]:
predictions = torch.argmax(probas, dim=1)
print(predictions)

tensor([0, 0, 0, 1, 1])


Note that it is unnecessary to compute softmax probabilities to obtain the class labels. Wecould also apply the
`argmax`
function to the logits (
`outputs`
) directly:

In [32]:
predictions = torch.argmax(outputs, dim=1)
print(predictions)

tensor([0, 0, 0, 1, 1])


Above, we computed the predicted labels for the training dataset. Since the training datasetis relatively small, we could compare it to the true training labels by eye and see that themodel is 100% correct. We can double-check this using the
`==`
comparison operator:

In [33]:
predictions == y_train

tensor([True, True, True, True, True])

In [34]:
torch.sum(predictions == y_train)

tensor(5)

However, to generalize the computation of the prediction accuracy, let’s implement a
`compute_accuracy`
function as shown in the following code.

In [35]:
def compute_accuracy(model, dataloader):

    model = model.eval()
    correct = 0.0
    total_examples = 0

    for idx, (features, labels) in enumerate(dataloader):

        with torch.no_grad():
            logits = model(features)

        predictions = torch.argmax(logits, dim=1)
        compare = labels == predictions
        correct += torch.sum(compare)
        total_examples += len(compare)

    return (correct / total_examples).item()

Note that the following
`compute_accuracy`
function iterates over a data loader to compute thenumber and fraction of the correct predictions. This is because when we work with large datasets, we typically can only call the model on a small part of the dataset due to memorylimitations. The
`compute_accuracy`
function above is a general method that scales to datasets of arbitrary size since, in each iteration, the dataset chunk that the model receives is the same size as the batch size seen during training.

Notice that the internals of the
`compute_accuracy`
function are similar to what we used beforewhen we converted the logits to the class labels.

In [36]:
compute_accuracy(model, train_loader)

1.0

In [37]:
compute_accuracy(model, test_loader)

1.0

## A.8 Saving and loading models

In the previous section, we successfully trained a model. Let’s now see how we can save atrained model to reuse it later.

Here’s the recommended way how we can save and load models in PyTorch:

In [38]:
torch.save(model.state_dict(), "model.pth")

The model’s
`state_dict`
is a Python dictionary object that maps each layer in the model to its trainable parameters (weights and biases).

Note that
`"model.pth"`
is an arbitrary filename for the model file saved to disk. We can give it any *name* and *file ending* we like; however,
`.pth`
and
`.pt`
are the most common conventions.

Once we saved the model, we can restore it from disk as follows:

In [39]:
model = NeuralNetwork(2, 2) # needs to match the original model exactly
model.load_state_dict(torch.load("model.pth", weights_only=True))

<All keys matched successfully>

The
`torch.load("model.pth")`
function reads the file
`"model.pth"`
and reconstructs the Python dictionary object containing the model’s parameters while
`model.load_state_dict()`
applies these parameters to the model, effectively restoring its learned state from when we saved it.

Note that the line
`model = NeuralNetwork(2, 2)`
above is not strictly necessary if you execute this code in the same session where you saved a model. However, I included it here to illustrate that we need an instance of the model in memory to apply the saved parameters. Here, the
`NeuralNetwork(2, 2)`
architecture needs to match the original saved model exactly.

## A.9 Optimizing training performance with GPUs

In this last section of this tutorial, we will see how we can utilize GPUs, which will accelerate deep neural network training compared to regular CPUs.

First, we will introduce the main concepts behind GPU computing in PyTorch.

Then, we will train a model on a single GPU.

Finally, we’ll then look at distributed training using multiple GPUs.

### A.9.1 PyTorch computations on GPU devices

As you will see, modifying the training loop from section 2.7 to optionally run on a GPU is relatively simple and only requires changing three lines of code.

Before we make the modifications, it’s crucial to understand the main concept behind GPU computations within PyTorch. First, we need to introduce the notion of devices. In PyTorch, a device is where computations occur, and data resides. The CPU and the GPU are examples of devices. A PyTorch tensor resides in a device, and its operations are executed on the same device.

Let’s see how this works in action. Assuming that you installed a GPU-compatible version of PyTorch as explained in section 2.1.3, Installing PyTorch, we can double-check that ourruntime indeed supports GPU computing via the following code:

In [40]:
import torch

print(torch.__version__)

2.8.0+cu126


In [41]:
print(torch.cuda.is_available())

False


Now, suppose we have two tensors that we can add as follows – this computation will be carried out on the CPU by default:

In [ ]:
tensor_1 = torch.tensor([1., 2., 3.])
tensor_2 = torch.tensor([4., 5., 6.])

print(tensor_1 + tensor_2)

tensor([5., 7., 9.])


We can now use the `.to()` method to transfer these tensors onto a GPU and perform theaddition there:

In [ ]:
tensor_1 = tensor_1.to("cuda")
tensor_2 = tensor_2.to("cuda")

print(tensor_1 + tensor_2)

tensor([5., 7., 9.], device='cuda:0')


Notice that the resulting tensor now includes the device information,
`device='cuda:0'`
, which means that the tensors reside on the first GPU. If your machine hosts multiple GPUs, you have the option to specify which GPU you’d like to transfer the tensors to. You can do this by indicating the device ID in the transfer command.

For instance, you can use
`.to("cuda:0")`
,
`.to("cuda:1")`
, and so on.

However, it is important to note that all tensors must be on the same device. Otherwise, the computation will fail, as shown below, where one tensor resides on the CPU and the other onthe GPU:

In [ ]:
tensor_1 = tensor_1.to("cpu")
print(tensor_1 + tensor_2)

RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

In this section, we learned that GPU computations on PyTorch are relatively straightforward.

All we have to do is transfer the tensors onto the same GPU device, and PyTorch will handle the rest.

Equipped with this information, we can now train the neural network from theprevious section on a GPU.

### A.9.2 Single-GPU training

Now that we are familiar with transferring tensors to the GPU, we can modify the training loop from section 2.7, A *typical training loop*, to run on a GPU.

In [ ]:
X_train = torch.tensor([
    [-1.2, 3.1],
    [-0.9, 2.9],
    [-0.5, 2.6],
    [2.3, -1.1],
    [2.7, -1.5]
])

y_train = torch.tensor([0, 0, 0, 1, 1])

X_test = torch.tensor([
    [-0.8, 2.8],
    [2.6, -1.6],
])

y_test = torch.tensor([0, 1])

In [ ]:
from torch.utils.data import Dataset


class ToyDataset(Dataset):
    def __init__(self, X, y):
        self.features = X
        self.labels = y

    def __getitem__(self, index):
        one_x = self.features[index]
        one_y = self.labels[index]
        return one_x, one_y

    def __len__(self):
        return self.labels.shape[0]

train_ds = ToyDataset(X_train, y_train)
test_ds = ToyDataset(X_test, y_test)

In [ ]:
from torch.utils.data import DataLoader

torch.manual_seed(123)

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=1,
    drop_last=True
)

test_loader = DataLoader(
    dataset=test_ds,
    batch_size=2,
    shuffle=False,
    num_workers=1
)

In [ ]:
class NeuralNetwork(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs):
        super().__init__()

        self.layers = torch.nn.Sequential(

            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30),
            torch.nn.ReLU(),

            # 2nd hidden layer
            torch.nn.Linear(30, 20),
            torch.nn.ReLU(),

            # output layer
            torch.nn.Linear(20, num_outputs),
        )

    def forward(self, x):
        logits = self.layers(x)
        return logits

**This requires only changing three lines of code, as shown in the code below.**

In [ ]:
import torch.nn.functional as F


torch.manual_seed(123)
model = NeuralNetwork(num_inputs=2, num_outputs=2)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu") # NEW
model.to(device) # NEW

# Note that the book originally used the following line, but the "model =" is redundant
# model = model.to(device) # NEW

optimizer = torch.optim.SGD(model.parameters(), lr=0.5)

num_epochs = 3

for epoch in range(num_epochs):

    model.train()
    for batch_idx, (features, labels) in enumerate(train_loader):

        features, labels = features.to(device), labels.to(device) # NEW
        logits = model(features)
        loss = F.cross_entropy(logits, labels) # Loss function

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        ### LOGGING
        print(f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
              f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
              f" | Train/Val Loss: {loss:.2f}")

    model.eval()
    # Optional model evaluation

Epoch: 001/003 | Batch 000/002 | Train/Val Loss: 0.75
Epoch: 001/003 | Batch 001/002 | Train/Val Loss: 0.65
Epoch: 002/003 | Batch 000/002 | Train/Val Loss: 0.44
Epoch: 002/003 | Batch 001/002 | Train/Val Loss: 0.13
Epoch: 003/003 | Batch 000/002 | Train/Val Loss: 0.03
Epoch: 003/003 | Batch 001/002 | Train/Val Loss: 0.00


In [ ]:
def compute_accuracy(model, dataloader, device):

    model = model.eval()
    correct = 0.0
    total_examples = 0

    for idx, (features, labels) in enumerate(dataloader):

        features, labels = features.to(device), labels.to(device) # New

        with torch.no_grad():
            logits = model(features)

        predictions = torch.argmax(logits, dim=1)
        compare = labels == predictions
        correct += torch.sum(compare)
        total_examples += len(compare)

    return (correct / total_examples).item()

In [ ]:
compute_accuracy(model, train_loader, device=device)

1.0

In [ ]:
compute_accuracy(model, test_loader, device=device)

1.0

In the case of the modified training loop above, we probably won’t see a speed-up becauseof the memory transfer cost from CPU to GPU. However, we can expect a significant speed-up when training deep neural networks, especially large language models.

As we saw in this section, training a model on a single GPU in PyTorch is relatively easy.Next, let’s introduce another concept: training models on multiple GPUs.

---------------------------------------------------------------------------

**IMPORTANT:**

**PyTorch on macOS**

On an Apple Mac with an Apple Silicon chip (like the M1, M2, M3, M4, or newermodels) instead of a computer with an Nvidia GPU, you can change
`device = torch.device("cuda"if torch.cuda.is_available() else "cpu")`
to
`device = torch.device("mps" iftorch.backends.mps.is_available() else "cpu")`
to take advantage of this chip.

-------------------------------------------------------


### A.9.3 Training with multiple GPUs

See [Multiple GPU DDP-Script.py](https://github.com/sahil-bhatia-iitbhu/pytorch/blob/40a486aefccf0adefb14510755637af55d3f086f/Pytorch%20in%20one%20hour%20v2/Pytorch%20in%20One%20Hour%20Multiple-GPU%20DDP-script.py)

<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/12.webp" width="600px">
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/13.webp" width="600px">